update with modern wrapper and style

k9cdt opened the pull request:
This PR comes with two major changes.

1. I rewrote the CLI with a typer, which provides a more modern and robust command line interface.
2. For a big part of the kernel, rewrote with numpy style and in a more pythonic way to enhance readability.

k9cdt committed:
update with modern wrapper and style

zwsmith200 commented on amino.py:
```diff
@@ -1,16 +1,16 @@
 """AMINO: generating a minimally redundant set of order parameters through
 clustering of mutual information based distances. Method by Ravindra, Smith,
-and Tiwary. Code maintained by Ravindra and Smith.
+and Tiwary. Code maintained by Ravindra and Smith. Tested and reorganized by 
+Da Teng in September 2024.
```
You can add yourself as a maintainer.

zwsmith200 commented on amino.py:
```diff
@@ -84,20 +84,22 @@ def d2_bin(self, x, y):
 
     # Checks if distance has been computed before, otherwise computes distance
-    def distance(self, OP1, OP2):
+    def distance(self, OP1: OrderParameter, OP2: OrderParameter) -> float:
```
:+1: Ideally we should have type hints everywhere, I'm in favor of adding them to at least every function you touch.

zwsmith200 commented on amino.py:
```diff
@@ -115,25 +117,49 @@ def distance(self, OP1, OP2):
+        float
+            The mutual information distance.
+            
+        """
```
Can we differentiate this from `distance` which should ideally have a more descriptive name? They currently have the same docstring.

zwsmith200 commented on amino_cli.py:
```diff
@@ -0,0 +1,45 @@
+    # Skip first column for time and read the rest of the file
+    time_series = np.loadtxt(filename).T[1:]
+
+    # Sanity check
+    assert(len(names) == time_series.shape[0])
```
Should this be an exception w/ more of a description? It would currently not be clear to a user why this happened (ie unexpectedly formatted input files).

k9cdt replied:
Yes. Sometimes PLUMED or whatever code that generated the time series can be problematic. I also want to ensure the headline was parsed correctly (no space in any OP names)

zwsmith200 commented on amino_cli.py:
```diff
@@ -0,0 +1,45 @@
+    if n is None:
+        n = len(names)
+    if n > 20 and not override:
+        n = 20
```
Should this have an associated warning? It seems like this could burn someone that didn't read the help message closely when we alter their input param silently.

zwsmith200 commented on amino_cli.py:
```diff
@@ -0,0 +1,45 @@
+def main(filename: Annotated[str, typer.Argument(help="Name of the COLVAR file containing the order parameters")], 
+         n: Annotated[int, typer.Option(help="Number of order parameters to be calculated")] = None, 
+         bins: Annotated[int, typer.Option(help="Number of bins")] = 50, 
+         kde_bandwidth: Annotated[float, typer.Option(help="Bandwidth for the KDE")] = 0.02,
```
`n` reads like a short name for an option. Should we have short/long forms like `-n` or `--num_ops` and `-b` or `--bins` for example?

zwsmith200 commented on README.md:
```diff
@@ -28,14 +24,19 @@ This directory contains the histogram-based Python 3 implementation of AMINO whi
+`<COLVAR>` is a mandatory argument providing the name of the COLVAR file that you want to reduce. 
 
-The default value for bins if the `-b` flag is omitted is 50.
+The `--n` option can be used to specify a maximum number of order parameters in the reduced COLVAR. Default value is 20 or the total number of order parameters in the provided <COLVAR> (whichever is smaller). AMINO will use a maximum number of output order parameters of 20, but you can use `--override` to remove this limitation.
```
What is the inspiration for this flag?

zwsmith200 commented on amino.py:
```diff
@@ -176,55 +202,57 @@ def add_OP(self, OP):
+                mut_info[i] = self.mut.distance(OP, op)
+            mut_info[n] = self.mut.distance(OP, OP)
+
+            try:
+                assert(mut_info[n] < 1e-2)
+            except:
+                raise ValueError("The dissimilarity between an OP and itself is not close to zero. Decrease bandwidth.")
```
:+1:

zwsmith200 replied:
I think bandwidth is no longer printed unless we’re in verbose mode. It’s a good idea to print it in this error message since it uses a rule of thumb by default and the user wouldn’t have a good idea where to start when manually setting BW.

zwsmith200 commented on amino.py:
```diff
@@ -176,55 +202,57 @@ def add_OP(self, OP):
+            mut_info = np.zeros(self.size)
+
+            n = len(self.OPs)
+            for i, op in enumerate(self.OPs):
```
Can we name this something more descriptive than `op` so we don't have `OP` and `op`. We should probably not use `OP` since caps usually implies constants.

k9cdt replied:
The use of the caped `OPs` was from the legacy code, which I didn't change. I can change the input variable `OP` to something else `new_op`

zwsmith200 commented on amino.py:
```diff
@@ -176,55 +202,57 @@ def add_OP(self, OP):
+
+        else: # adding an OP when there are fewer than self.size
+
+            mut_info = np.zeros(self.size)
+
+            n = len(self.OPs)
+            for i, op in enumerate(self.OPs):
+                mut_info[i] = self.mut.distance(OP, op)
+            mut_info[n] = self.mut.distance(OP, OP)
+
+            try:
+                assert(mut_info[n] < 1e-2)
+            except:
+                raise ValueError("The dissimilarity between an OP and itself is not close to zero. Decrease bandwidth.")
+
+            self.matrix[n,:] = mut_info
+            self.matrix[:,n] = mut_info
```
I think we can actually clean this up further since there is a lot of repeated functionality like adding `mut_info` to `matrix` and the `self.mut.distance` loop.

Maybe we should also write a unit test for this before we change it since it's becoming a more complicated refactor.

zwsmith200 commented on amino.py:
```diff
@@ -360,29 +326,49 @@ def cluster(ops, seeds, mut):
+            my_group = np.nonzero(group == i)[0]
 
-    return centers
+            dtt = [np.sum([mut.distance(ops[g], ops[j]) ** 2 for j in my_group]) for g in my_group]
```
Can this have a more descriptive name?

zwsmith200 commented on amino.py:
```diff
@@ -327,19 +292,20 @@ def get_matrix(ops, mut):
 
 # Running clustering on `ops` starting with `seeds`
-def cluster(ops, seeds, mut):
+def cluster(ops: list[OrderParameter], seeds: list[OrderParameter], mut: Memoizer) -> list[OrderParameter]:
```
I like this change, should this also get a unit test on a small synthetic data set to make sure we aren't breaking anything?

zwsmith200 reviewed (commented):
Thanks for going through this. I like that loops are more in line with python/numpy best practices.

k9cdt committed:
addressed PR #7 comments

k9cdt commented:
I addressed most comments with a new commit except for the `add_OP` function. The `if` statement makes it hard to consolidate those matrix constructions without introducing a new function. 

I agree with adding more tests. Maybe some synthetic data for sanity checks.

k9cdt committed:
changed add_OP to recap old code behavior
